FIRMWARE.md

```diff
@@ -0,0 +1,252 @@
+# Firmware notes
+
+## Early July update
+
+Starting early July 2020, this project provides a **beta** quality
+graphics-and-USB enabled U-Boot for the Pinebook Pro. This is made possible by
+using a preview of upcoming patches to U-Boot. See the *Acknowledgements*
+section of this document.
+
+This will be called the **opinionated firmware**. It deviates from some U-Boot
+defaults, while keeping in the spirit of the original defaults.
+
+> **TIP**: Look at the *Known issues* section to know about the known issues!
+
+This U-Boot build should be usable by users from other distributions, though
+has been customized a bit visually for NixOS users.
+
+The boot order is, with the default build, the upstream boot order from U-Boot.
+Which means that it will first try to boot from eMMC, then from SD, then from
+USB.
+
+Our opinionated changes, though, make it trivial to select an alternative boot
+option for one boot. You can cancel the default boot by mashing CTRL+C when
+prompted (and a bit later too).
+
+Once canceled, you will be presented with a menu depicting the boot options
+that are normally tried in order.
+
+```
+
+  *** U-Boot Boot Menu ***
+
+     Default U-Boot boot
+     Boot from eMMC
+  => [Boot from SD]
+     Boot from USB
+     Boot PXE
+     Boot DHCP
+     Reboot
+     U-Boot console
+
+
+  Press UP/DOWN to move, ENTER to select
+
+```
+
+By selecting another option, the firmware will try to boot once from the given
+storage using the default *U-Boot distro commands*.
+
+At the time of writing, the default boot sources on a device are:
+
+ * Bootable partition with an extlinux.conf compatible boot under `/` or `/boot` (`scan_dev_for_extlinux`)
+ * Bootable partition with boot scripts `boot.scr.uimg` or `boot.scr` under `/` or `/boot` (`scan_dev_for_scripts`)
+ * UEFI program `efi/boot/bootaa64.efi` (`scan_dev_for_efi`)
+
+### Building
+
+If you're building on x86_64, it will automatically use cross-compilation. In
+that case the build should work assuming cross-compilation is working fine on
+current unstable.
+
+Otherwise, on-device or on any other AArch64 system, the build should be
+trivial.
+
+```
+ $ nix-build -I nixpkgs=channel:nixos-unstable -A pkgs.pinebookpro-firmware-installer
+```
+
+### Installing to SPI flash
+
+This is not mandatory. The opinionated firmware still works just as well
+installed to SD or to eMMC.
+
+These instructions assume the use of the opinionated U-Boot we are providing.
+You can also install to SPI using any other documented method.
+
+Before starting, you will need an SD card on which you can write the Firmware
+installer image. This will overwrite the contents of that SD card.
+
+```
+ # Replace sdX with the correct block device for your SD card.
+ $ sudo dd if=firmware-installer-image.img of=/dev/sdX bs=512 seek=64 oflag=direct,sync
```

Comment:
```suggestion
 $ sudo dd if=firmware-installer-image.img of=/dev/sdX bs=512 oflag=direct,sync
```

Removing this fixed the issue for me.  I presume this is a copy/paste issue? Otherwise the partition table is not even properly aligned.

With this I was able to erase the SPI flash, reboot, rewrite the SPI flash, and finally reboot into Linux.

Reply:
It might also be worthwhile to add a note that the SD card is `/dev/disk/by-path/platform-fe320000.dwmmc` or `/dev/mmcblk1` on the PBP, in case (like for me), the user wants to use an already booted PBP to write the SD card.

Reply:
Oops. Yes. That is it.